### ISSUE TITLE
K8SSAND-580 ⁃ Migrate reaper_controller to k8ssandra-operator

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#4]
[reaper-operator](https://github.com/k8ssandra/reaper-operator) consists of the Reaper CRD, the reaper_controller, and the cassandradatacenter_controller. reaper_controller configures and deploys Reaper. We want migrate that code and the Reaper CRD into this repo. It will probably be easiest to recreate the CRD. We just need to make sure it has the same fields. 

cassandradatacenter_controller's main function is registering CassandraDatacenters with Reaper. We can move this functionality into k8ssandracluster_controller.

For this ticket assume that a Reaper instance per CassandraDatacenter is the default.

Integration and e2e tests should be added/updated as necessary.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-580) by [Unito](https://www.unito.io)
┆Epic: Reaper Controller Integration
┆Issue Number: K8SSAND-580
┆Priority: Medium

**Comment by adutra** [comment]
I created a draft PR here: https://github.com/k8ssandra/k8ssandra-operator/pull/216.

I migrated the API, controllers and pkg content to k8ssandra-operator and modified a few things accordingly (API group for instance).

> cassandradatacenter_controller's main function is registering CassandraDatacenters with Reaper. We can move this functionality into k8ssandracluster_controller.

@jsanda I suppose we want to adopt for Reaper the same design principles in place for Stargate? That is, a global template at k8c level, overridable at DC level?

Also, the current way Reaper instances are wired to DCs is through a specific annotation placed _in the target cassdc resource_: `reaper.k8ssandra.io/instance`. Do we want to keep honoring this mechanism? Because this is the opposite of what we do to connect stargates to cassdcs: the stargate resource has a reference to the DC name (but the target cassdc resource itself is unaware of the stargate resource).

**Comment by jsanda** [comment]
Agreed that we don't need the cassandaradatacenter_controller and yeah, I think it makes sense to follow the same design with respect to configuring at cluster and DC level. One difference though is I don't think we need to go beyond the DC level with Reaper whereas with Stargate it can be configured at the rack level.

**Comment by adutra** [comment]
I am making good progress but also facing some challenges. I will try to summarize everything below:

### Overall migration

Action items:
- [x] reaper-controller moved to new API group + controller
- [x] datacenter-controller logic incorporated in k8ssandra-controller
- [x] migrate `reaper` package and subpackages + unit tests. Subpackages merged into one single `reaper` package. 
- [ ] migrate controller tests, remove Ginkgo.
- [ ] migrate e2e tests, remove Ginkgo.

### Introduce a `ReaperTemplate` type

Action items:
- [x] Subset of `ReaperSpec` that can appear at cluster-level and DC level
- [x] Same "coalesce" semantics already in place for `K8ssandraCluster` and `Stargate`

### Connecting `CassandraDatacenter` and `Reaper` instances

Currently, the user puts an annotation on `CassandraDatacenter` resource. But we usually do the opposite: the `Reaper` instance should "point" to `CassandraDatacenter` (cf. Stargate). Although, I must confess the annotation approach has some advantages, e.g. a DC cannot reference more than one associated resource. @jsanda do you have any preference? If not, I'm going to go with a Stargate-like approach.

Action items: 
- [x] user or k8ssandra-controller assigns reference field to `CassandraDatacenter` resource on `ReaperSpec`; field is mandatory (same as Stargate).
  - This field is DISTINCT from CassandraBackend ref, see next topic.

### Confusion between target DC and backend DC

This came as a surprise for me: when creating new deployments reaper-controller looks up the backend DC, not the target DC. See [here](https://github.com/k8ssandra/reaper-operator/blob/637cb522d19fb106deef465408f175435c56aae5/pkg/reconcile/reconcilers.go#L189-L201). I don't explain how this can work when using in-memory storage or different target/backend DCs. @adejanovski do you know? Am I missing something obvious?

Action items:
- [x] Change DC-lookup logic to lookup the DC based on the reference field on `ReaperSpec`. DO NOT use the backend DC ref for that.
- [ ] Change validation of backend DC: if storage type is Cassandra and no backend provided, use the target DC instead of returning a validation error.

### Schema reconciliation

Currently: done in 2 places: 
- during reconciliation of a Reaper instance: keyspace creation
- inside an init-container in each pod in the Reaper deployment: table upgrades (it seems)

IMO this can only work because Reaper deployments have 1 single replica, and as long as user does not create 2 Reaper resources in the same target DC. Otherwise we'd run into the same issue as for Stargate (concurrent schema updates). @adejanovski do you confirm?

Action items:
- [ ] Schema reconcile should be done at `K8ssandraCluster` level IIF: 
  - the `Reaper` instance was created by that controller (same as Stargate auth schema);
  - the `Reaper` instance is configured with Cassandra backend.
- [ ] Otherwise, if the `Reaper` resource was created manually: do nothing and expect user to have created the schema upfront, if required.

### Schema replication

Currently users select the desired replication in a `ReplicationConfig` object. I wonder if this should be kept? Seems like a nice way for users to shoot themselves in the foot. Indeed if the `Reaper` instance was created by k8ssandra-controller, then the controller knows the best replication, especially in multi-DC setups (see next topic). If not, then user is expected to have created the keyspace with appropriate replication.

Action items:
- [ ] Remove `ReplicationConfig`;
- [ ] K8ssandra controller should compute the desired replication, if necessary.

### Multi-DC setups

Reading [this page](https://cassandra-reaper.io/docs/usage/multi_dc/), in a multi-DC setup with Cassandra backend I think we need to adopt the approach named "Multiple Reaper instances with JMX accessible for the local DC only":

- The `reaper` keyspace would be created only once by k8ssandra-controller;
- The _contents_ of the keyspace would be created by the init-container in the first Reaper deployment in the first DC. Since reapers are deployed sequentially, this should be OK.

@adejanovski can you please just confirm for me: when using `datacenterAvailability: EACH` do all Reaper instances share the same `reaper` keyspace?

**Comment by adejanovski** [comment]
> This came as a surprise for me: when creating new deployments reaper-controller looks up the backend DC, not the target DC. See here. I don't explain how this can work when using in-memory storage or different target/backend DCs. @adejanovski do you know? Am I missing something obvious?

I think in-memory should never be used unless it's for testing.
While reaper-operator could support repairing multiple clusters, thus distinguishing the backend DC and the target DC, in the context of k8ssandra-operator I would say both should be the same. We're repairing a single Cassandra cluster and are using it as well as backend.

> IMO this can only work because Reaper deployments have 1 single replica, and as long as user does not create 2 Reaper resources in the same target DC. Otherwise we'd run into the same issue as for Stargate (concurrent schema updates). @adejanovski do you confirm?

The schema migration library we use performs leader election prior to doing the migrations to ensure we don't have concurrent schema updates.

> @adejanovski can you please just confirm for me: when using datacenterAvailability: EACH do all Reaper instances share the same reaper keyspace?

Yes, when using `EACH`, the `reaper` keyspace (usually named `reaper_db`) will be used by all Reaper instances as a way of coordinating work and exchanging metrics.

**Comment by adutra** [comment]
> I think in-memory should never be used unless it's for testing. While reaper-operator could support repairing multiple clusters, thus distinguishing the backend DC and the target DC, in the context of k8ssandra-operator I would say both should be the same. We're repairing a single Cassandra cluster and are using it as well as backend.

Excellent, if you agree then I will remove the in-memory storage type. This will greatly simplify the new CRD.

**Comment by adejanovski** [comment]
Please do, it's useless in our context 👍 
Also, Postgres and H2 are both gone from Reaper 3.0.0, so it's just Cassandra and Astra now (not sure that matters for this task though).

**Comment by jsanda** [comment]
Catching up on the ticket...

I am in favor of removing the memory backend, even if it means updating some tests.

With respect to target DC and backend, I am in favor of the simplification. The annotation isn't necessary since the K8ssandraCluster controller is managing the orchestration. In k8ssandra 1.x we don't have that orchestration layer.

**Comment by jsanda** [comment]
I took a first pass reviewing #216 and want to provide more feedback on some of the things covered in https://github.com/k8ssandra/k8ssandra-operator/issues/4#issuecomment-962972696. (Note that comments to follow are based on code in #216)

**Orchestration**
First, I want to discuss the sequence of events for the orchestration performed by K8ssandraClusterReconciler. It first reconciles all CassandraDatacenters. It will continue to requeue until all CassandraDatacenters are ready. 

Once all CassandraDatacenters are ready it proceeds to reconcile Stargate for each DC in a similar fashion. Suppose we have three DCs - dc1, dc2, dc3 - with Stargate to be deployed in each DC. K8ssandraClusterReconciler will first deploy Stargate in dc1. It will deploy Stargate in dc2 only after Stargate in dc1 is ready and then deploy in dc3 only after Stargate in dc2 is ready. For the first Stargate instance, K8ssandraClusterReconciler reconciles the Stargate auth schema. After the auth schema is reconciled, the Stargate instances _can_ be deployed all at once. We don't need to wait until Stargate in dc1 is ready before moving on to deploy Stargate in dc2 and dc3. We have a similar situation with Reaper.

Continuing with the previous example, suppose we are deploying Reaper in dc1, dc2, and in dc3. K8ssandraClusterReconciler will deploy Reaper in dc1 only after all Stargates are ready. The Reaper schema is initialized when Reaper is deployed in dc1. Reaper is deployed in dc2 and in dc3 only after Reaper is ready in dc1 and in dc2 respectively. Let's assume for the moment that we can initialize the Reaper schema separately before deploying any Reaper instances. As long as the schema is initialized first, we can and should deploy Reaper across all DCs at the same time (i.e., within the same reconciliation). Furthermore, we can and should deploy Reaper and Stargate at the same time.

I am not suggesting that the changes happen in #216. In fact, I'd say they are outside the scope of this issue; however they absolutely should happen. One of the reasons I wanted to share my thoughts on the orchestration is that it impacts some of the other things mentioned, namely schema reconciliation.

**schema reconciliation**
Like the Stargate auth keyspace, the Reaper keyspace needs to be initialized by K8ssandraClusterReconciler. This includes configuring replication. Unlike Stargate, Reaper schema initialization is much more involved as it configures/updates all the tables required by Reaper. If we continue to rely on initializing schema with an init container, then we would need to create the first Reaper instance and wait until it is ready before deploying other Reaper instances in other DCs. This however is suboptimal. We should initialize the schema up front and then deploy all Reaper instances at once (this is assuming all CassandraDatacenters are already ready).

I absolutely do not want to implement Reaper schema initialization in the operator :) Reaper already uses a library to handle new installs and upgrades. This should be wrapped in a k8s job. We wouldn't even need to worry about leader election. Again, this would be fall outside the scope of this issue IMO.

**Comment by adutra** [comment]
Thanks @jsanda ! 

> After the auth schema is reconciled, the Stargate instances can be deployed all at once. We don't need to wait until Stargate in dc1 is ready before moving on to deploy Stargate in dc2 and dc3. We have a similar situation with Reaper.

Agreed, Stargates and Reaper should be deployed all at once, _after_ their schemas are reconciled. I also agree that this should be done outside of this ticket, because I think we need now to introduce a helper component, something similar to `ReconciliationContext` in cass-operator, to make these changes easier to implement.

> We should initialize the schema up front and then deploy all Reaper instances at once (this is assuming all CassandraDatacenters are already ready). I absolutely do not want to implement Reaper schema initialization in the operator :) Reaper already uses a library to handle new installs and upgrades. This should be wrapped in a k8s job.

Agreed again, but one question: what about people using Reaper as standalone instances? Is our stance always the same here? – that is: users manually creating Stargates and Reapers are also responsible for their schemas.

One last thing: you've noticed that my PR is based on #156 unfortunately. There are bits of code there that I needed for Reaper. Once #156 is merged I will cleanup this PR.

**Comment by jsanda** [comment]
>what about people using Reaper as standalone instances? Is our stance always the same here? – that is: users manually creating Stargates and Reapers are also responsible for their schemas.

I think that is a fair stance. Whatever (or whoever) is managing the orchestration should configure the schemas.
